File: misc/cpp/chunking/MimeUtils.cpp
/*
 * MimeUtils.cpp
 *
 * This file is part of the IHMC Misc Library
 * Copyright (c) 2010-2016 IHMC.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 3 (GPLv3) as published by the Free Software Foundation.
 *
 * U.S. Government agencies and organizations may redistribute
 * and/or modify this program under terms equivalent to
 * "Government Purpose Rights" as defined by DFARS 
 * 252.227-7014(a)(12) (February 2014).
 *
 * Alternative licenses that allow for use within commercial products may be
 * available. Contact Niranjan Suri at IHMC (nsuri@ihmc.us) for details.
 */

#include "MimeUtils.h"

#include "Defs.h"
#include "File.h"
#include "FileUtils.h"
#include "Logger.h"
#include "NLFLib.h"

#include <string.h>

#ifdef WIN32
    #define stringcasecmp _stricmp
#else
    #define stringcasecmp strcasecmp
#endif

using namespace IHMC_MISC;
using namespace NOMADSUtil;

const char * MimeUtils::DEFAULT_MIME_TYPE = "application/octet-stream";

NOMADSUtil::String MimeUtils::getMimeType (const char *pszFileName)
{
    if (pszFileName == NULL) {
        checkAndLogMsg ("MimeUtils::getMimeType", Logger::L_Warning,
                        "File name is null\n");
        return String();
    }
    if (!FileUtils::fileExists (pszFileName)) {
        checkAndLogMsg ("MimeUtils::getMimeType", Logger::L_Warning,
                        "File %s not found\n", pszFileName);
        return String();
    }
    File file (pszFileName);
    return toType (file.getExtension());
}

Chunker::Type MimeUtils::mimeTypeToFragmentType (const char *pszMimeType)
{
    if (pszMimeType == NULL) {
        return Chunker::UNSUPPORTED;
    }
    if ((0 == stringcasecmp (pszMimeType, "image/x-ms-bmp")) ||
        (0 == stringcasecmp (pszMimeType, "image/x-bmp"))) {    
        return Chunker::BMP;
    }
    if (0 == stringcasecmp (pszMimeType, "image/jpeg")) {
        return Chunker::JPEG;
    }
    if (0 == stringcasecmp (pszMimeType, "image/jp2")) {
        return Chunker::JPEG2000;
    }
    if (0 == stringcasecmp (pszMimeType, "video/mpeg")) {
        return Chunker::V_MPEG;                     
    }
    if (0 == stringcasecmp (pszMimeType, "audio/mpeg")) {
        return Chunker::A_MPEG;
    }
    if (0 == stringcasecmp (pszMimeType, "image/png")) {
        return Chunker::PNG;
    }
    if ((0 == stringcasecmp (pszMimeType, "application/x-troff-msvideo")) ||
        (0 == stringcasecmp (pszMimeType, "video/avi")) ||
        (0 == stringcasecmp (pszMimeType, "video/msvideo")) ||
        (0 == stringcasecmp (pszMimeType, "video/x-msvideo"))) {
        return Chunker::AVI;
    }
    if (0 == stringcasecmp (pszMimeType, "video/quicktime")) {
        return Chunker::MOV;
    }
    if (0 == stringcasecmp (pszMimeType, "video/mpeg")) {
        return Chunker::V_MPEG;
    }

    return Chunker::UNSUPPORTED;

}

Chunker::Type MimeUtils::toType (const String &extension)
{
    if (extension ^= "bmp") {
        return Chunker::BMP;
    }
    if ((extension ^= "jpeg") || (extension == "jpg")) {
        return Chunker::JPEG;
    }
    if (extension ^= "jp2") {
        return Chunker::JPEG2000;
    }
    if (extension ^= "png") {
        return Chunker::PNG;
    }
    if (extension ^= "avi") {
        return Chunker::AVI;
    }
    if (extension ^= "mov") {
        return Chunker::MOV;
    }
    if (extension ^= "mp4") {
        return Chunker::V_MPEG;
    }
    if (extension ^= "mpg") {
        return Chunker::V_MPEG;
    }
    return Chunker::UNSUPPORTED;
}

String MimeUtils::toExtesion (Chunker::Type extension)
{
    switch (extension) {
        case Chunker::BMP: return String ("bmp");
        case Chunker::JPEG: return String ("jpg");
        case Chunker::JPEG2000: return String ("jp2");
        case Chunker::PNG: return String ("png");
        case Chunker::AVI: return String ("avi");
        case Chunker::MOV: return String ("mov");
        case Chunker::V_MPEG: return String ("mpeg");
        default: return String();
    }
}

String MimeUtils::toMimeType (Chunker::Type extension)
{
    switch (extension) {
        case Chunker::BMP: return String ("image /x-bmp");
        case Chunker::JPEG: return String ("image/jpeg");
        case Chunker::JPEG2000: return String ("image/jp2");
        case Chunker::PNG: return String ("image/png");
        case Chunker::AVI: return String ("video/avi");
        case Chunker::MOV: return String ("video/quicktime");
        case Chunker::V_MPEG: return String ("video/mpeg");
        default: return String();
    }
}

File: util/java/us/ihmc/util/MIMEUtils.java
/*
 * MIMEUtils.java
 *
 * This file is part of the IHMC Util Library
 * Copyright (c) 1993-2016 IHMC.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 3 (GPLv3) as published by the Free Software Foundation.
 *
 * U.S. Government agencies and organizations may redistribute
 * and/or modify this program under terms equivalent to
 * "Government Purpose Rights" as defined by DFARS 
 * 252.227-7014(a)(12) (February 2014).
 *
 * Alternative licenses that allow for use within commercial products may be
 * available. Contact Niranjan Suri at IHMC (nsuri@ihmc.us) for details.
 */

package us.ihmc.util;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Enumeration;
import java.util.HashMap;
import java.util.ResourceBundle;
import java.util.regex.Pattern;
import javax.activation.MimetypesFileTypeMap;

/**
 *
 * @author Giacomo Benincasa        (gbenincasa@ihmc.us)
 */
public class MIMEUtils
{
    public static final Pattern MSDOC = Pattern.compile ("application/.*\\.openxmlformats-officedocument\\..*");
    public static final Pattern MSDOC_OLD = Pattern.compile ("application/.*(\\.)*((.*excel)|(.*msword)|(.*powerpoint))");
    public static final Pattern OPENDOC = Pattern.compile ("application/.*\\.opendocument\\..*");

    private static final String MIME_TYPE_FILE = "/us/ihmc/util/mime/MIMETypes.properties";
    private static final String MIME_TYPE_FILEP = "us.ihmc.util.mime.MIMETypes";

    // File extension -> MIME Type
    // Load custom mime-types file (since its location appears not to be that standard...
    private static final MimetypesFileTypeMap _mimeTypes = new MimetypesFileTypeMap (MIMEUtils.class.getResourceAsStream (MIME_TYPE_FILE));

    // MIME Type -> File extension
    private static final HashMap<String, String> _fileExtensions = new HashMap<String, String>();
    static {
        ResourceBundle bundle = ResourceBundle.getBundle (MIME_TYPE_FILEP);
        for (Enumeration e = bundle.getKeys(); e.hasMoreElements(); ) {
            String type = (String) e.nextElement();
            String[] extensions = (bundle.getString (type)).split ("s+");
            if (_fileExtensions.get (type) == null) {
                _fileExtensions.put (type, extensions[0]);
            }
        }
    }

    public static String getMIMEType (File file) throws Exception
    {
        if (file.exists()) {
            String type = _mimeTypes.getContentType (file);
            if (type == null) {
                throw new Exception ("No MIME type for file " + file.getAbsolutePath());
            }
            return type;
        }
        throw new FileNotFoundException ("File " + file.getAbsolutePath() + " not found");
    }

    /**
     * Given the MIME type, it returns a possible file extension (or null if a
     * mapping was not found)
     */
    public static String getExtension (String mimeType)
    {
        if (mimeType == null || mimeType.length() == 0) {
            return null;
        }
        return _fileExtensions.get (mimeType);
    }

    /**
     * Return true if the MIMI type matches MSDOC or MSDOC_OLD or OPENDOC
     */
    public static boolean isDocumentType (String mimeType)
    {
        if (MSDOC.matcher(mimeType).matches() ||
            MSDOC_OLD.matcher(mimeType).matches() ||
            OPENDOC.matcher(mimeType).matches()) {
            return true;
        }
        return false;
    }

    public static boolean isText (String mimeType)
    {
        if (mimeType.contains("wordprocessing") ||
            mimeType.contains("opendocument.text") ||
            mimeType.contains("msword")) {
            return true;
        }
        return false;
    }

    public static boolean isSpreadsheet (String mimeType)
    {
        if (mimeType.endsWith (".sheet") ||
            mimeType.contains("spreadsheet") ||
            mimeType.contains("excel")) {
            return true;
        }
        return false;
    }

    public static boolean isPresentation (String mimeType)
    {
        if (mimeType.contains("presentation") ||
            mimeType.contains("ms-powerpoint") ||
            mimeType.contains("slideshow")) {
            return true;
        }
        return false;
    }

    public static boolean isImage (String mimeType)
    {
        if (mimeType.contains("image") ||
            mimeType.contains("IMAGE")) {
            return true;
        }
        return false;
    }

    public static void main (String[] args) throws Exception
    {
        for (String file : args) {
            System.out.println ("The MIME type of file " + file + " is: " + getMIMEType (new File (file)));
        }
        ArrayList<String> strings = new ArrayList<String>();

        strings.add ("application/vnd.openxmlformats-officedocument.wordprocessingml.document");
        strings.add ("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet");
        strings.add ("application/vnd.openxmlformats-officedocument.presentationml.presentation");
        strings.add ("application/vnd.openxmlformats-officedocument.presentationml.slideshow");
        for (String string : strings) {
            if (!MSDOC.matcher(string).matches()) {
                System.out.println (string + " does not match" + " MSDOC");
            }
        }

        strings.clear();
        strings.add ("application/msword");
        strings.add ("application/vnd.ms-excel");
        strings.add ("application/vnd.ms-powerpoint");
        for (String string : strings) {
            if (!MSDOC_OLD.matcher(string).matches()) {
                System.out.println (string + " does not match" + " MSDOC_OLD");
            }
        }

        strings.clear();
        strings.add ("application/vnd.oasis.opendocument.text");
        strings.add ("application/vnd.oasis.opendocument.spreadsheet");
        strings.add ("application/vnd.oasis.opendocument.presentation");
        for (String string : strings) {
            if (!OPENDOC.matcher(string).matches()) {
                System.out.println (string + " does not match" + " OPENDOC");
            }
        }

    }
}
